### ISSUE TITLE
Performance bottleneck example (Yii2 ActiveRecord)

### ISSUE THREAD

**Issue by przepompownia** [phpactor/phpactor#1125]
After experiencing some performance problems with phpactor I created [an example project](https://github.com/przepompownia/example-legacy-project) to demonstrate some of them. The most of this project comes from [composer create-project --prefer-dist yiisoft/yii2-app-basic basic](https://www.yiiframework.com/doc/guide/2.0/en/start-installation). Notice that the PSR-4 autoloader was added manually. 

Reproduction:
- install the project
- uncomment [this line](https://github.com/przepompownia/example-legacy-project/blob/9c90bad806cf2787750c8ebe6110ff3e130d69c8/yii2-basic/modules/example/commands/StupidController.php#L23) 
- get completions using RPC (with `rpc.store_replay: true`)
- use some profiler to replay the latest RPC request (e.g. `blackfire run phpactor rpc --replay`)

For some reason I could not get completions using RPC (without mute notices in https://github.com/phpactor/worse-reflection/blob/682bb4197ad014f10988dc1f1e2335a36902fd34/lib/Core/Inference/SymbolContextResolver.php#L236) while the same notices were ignored by LSP client. Probably we already discussed somewhere about mixing stdin with stderr in RPC output.

https://blackfire.io/profiles/5e7aa65d-886a-4394-98fa-30d8126d3c5e/graph - I am not sure how long this profile will be available.
```
Wall Time  1min 30s
Memory       13.4MB
```

At this moment I would like to show this case without investigating how it could be optimized.

**Comment by dantleech** [comment]
So most of the time seems to come from scanning the FS - it looks like it's using the fallback "simple" or brute-force class name to file converter which is a performance disaster (it scans all files in the project). This can be disabled and is disabled by default with the language server:

https://github.com/phpactor/phpactor/blob/6ee1603677757a1a9d42dfbd9f924347d6196a3e/lib/Phpactor.php#L259-L265

Do you get the same performance issues with the LS?

**Comment by przepompownia** [comment]
> Do you get the same performance issues with the LS?

Yes: I used RPC to simplify the reproduction only. I need some time to verify it, especially find the issue that shows how to replay LSP scenarios (reduced to needed requests, especially when we already have lots of `textDocument/documentHighlight` requests). It seems that we have documented replaying only for RPC calls.

LSP requests are asynchronous so it looked like no response after jumping to another place. The log shows that the answers came far later.

**Comment by dantleech** [comment]
Profiling the language server has been a tricky problem - ideally I'd like to have a way to enable / disable profiling while it is running.

But at least in the documented case above disabling "BRUTE_FORCE_CONVERSION" should help.

**Comment by przepompownia** [comment]
Independently of experimenting with `BRUTE_FORCE_CONVERSION` in free time I would like to understand what information are still needed to be gathered with each call from scratch or what is persisted in the index. Superficial reading of the log suggests that, for example, `WorkspaceSourceLocator` does not persist gathered data anywhere. 

```
[2020-11-08 19:26:52] phpactor.DEBUG: NOK [0.0000] "pcntl_sigtimedwait" with locator "Phpactor\Extension\LanguageServerWorseReflection\SourceLocator\WorkspaceSourceLocator" : Class "pcntl_sigtimedwait" not found [] []
[2020-11-08 19:26:52] phpactor.DEBUG: NOK [0.0000] "pcntl_sigtimedwait" with locator "Phpactor\Indexer\Adapter\Worse\IndexerClassSourceLocator" : Class "pcntl_sigtimedwait" is indexed, but it does not exist at path ""! [] []
[2020-11-08 19:26:52] phpactor.DEBUG:  OK [0.0000] "pcntl_sigtimedwait" with locator "Phpactor\Indexer\Adapter\Worse\IndexerFunctionSourceLocator" [] []
```

**Comment by dantleech** [comment]
`WorkspaceSourceLocator` is an in-memory locator which is used by the language server - it will keep track of classes / functions in open files which may not been persisted to disk.

**Comment by przepompownia** [comment]
The [option mentioned above](https://github.com/phpactor/phpactor/issues/904#issuecomment-596395393
) was removed with https://github.com/phpactor/language-server-extension/commit/ebdd15f755538b3fbd2a55460627d85a0311940d.

Blackfire has SDK that [allows using its probe on a part of code](https://blackfire.io/docs/integrations/php/sdk).

**Comment by przepompownia** [comment]
>  ideally I'd like to have a way to enable / disable profiling while it is running (which I think is possible with Blackfire at least)

This simplest usage works on [`complete()`](https://github.com/phpactor/completion/blob/master/lib/Core/ChainCompletor.php#L28): 
```php
    public function complete(TextDocument $source, ByteOffset $offset): Generator
    {
        $isComplete = true;

        $blackfire = new \Blackfire\Client();
        $probe = $blackfire->createProbe();
        foreach ($this->completors as $completor) {
            $suggestions = $completor->complete($source, $offset);

            yield from $suggestions;

            $isComplete = $isComplete && $suggestions->getReturn();
        }
        $profile = $blackfire->endProbe($probe);

        return $isComplete;
    }
```

**Comment by dantleech** [comment]
I'd quite like an custom command which starts blackfire, e.g. `:CocCommand phpactor.blackfire_start_trace` (or less intrusively `:CocRequest('phpactor', 'blackfire/start')`

but yeah, hard coding the probe is also fine for now :)

**Comment by elythyr** [comment]
> So most of the time seems to come from scanning the FS - it looks like it's using the fallback "simple" or brute-force class name to file converter which is a performance disaster (it scans all files in the project). This can be disabled and is disabled by default with the language server:
> 
> https://github.com/phpactor/phpactor/blob/6ee1603677757a1a9d42dfbd9f924347d6196a3e/lib/Phpactor.php#L259-L265
> 
> Do you get the same performance issues with the LS?

Why only disable them for LSP ?
I thought the indexer was now used with RPC too since the integration of a "watcher agent".
It might be handy to try to identify "legacy" code like this one that we might be able to remove in favor of the indexer.
We could then start to focus a bit more maybe on trying to deprecate things, identifying which one are almost not used anymore except for one or two spots.
This will be some nice technical backlog tickets helping us to sanitize the code while migrating towards a domain that can be shared between LSP & RPC.

I feel like I mostly try to do PR to add new features, even small ones, because I don't have a good enough view of the all project.
I personally pretty please with what phpactor as to offer and even if we still have a lot of things we could add, I think we might want to slow down the features a bit to take care of some legacy issue and maybe reduce a bit the issue counter :see_no_evil:

**Comment by dantleech** [comment]
> I thought the indexer was now used with RPC too since the integration of a "watcher agent".

RPC does not support background indexing or watching (although you can do it via. the CLI) and this is a heavy process more suited to a language server - there is still a case for RPC being lightweight (although finally we may depend on the indexer more and more).

> We could then start to focus a bit more maybe on trying to deprecate things

It might be good to start a new project `phpactor-ls` which is only a language server, then the two projects can be separate (but share the same components) - although this might just make maintenance harder again.

In any case, having a separate `phpactor-ls` project would force focus on matching the RPCs capabilities, the problem is always time however ...

**Comment by przepompownia** [comment]
Thanks to looking at `ProfilerMiddleware` and my example above I created simple wrapper to local usage at the moment:
https://github.com/phpactor/phpactor/compare/master...przepompownia:blackfire-middleware?expand=1

Example result: completion on `$trash->` from https://github.com/przepompownia/example-legacy-project/blob/master/yii2-basic/modules/example/commands/StupidController.php
https://blackfire.io/profiles/2fd1bd74-9f47-4041-b790-a3038c993188/graph

**Comment by dantleech** [comment]
Assume this i using the LSP and not RPC like the original issue.

Looks like the DocblockParser is getting hit quite alot, are the YII classes annotation heavy? I wonder if simple memoization would help.

**Comment by przepompownia** [comment]
Yes, on this branch it is currently hardcoded into `LanguageServerExtension`.

In https://github.com/przepompownia/example-legacy-project you can see five levels of inheritance over `Trash` (not counting interfaces). Each of them, including `Trash`, has a lot of docblocks. Currently I don't know how much they all have an impact for finding suggestions. It's typical for Yii framework classes. I encourage you to clone it (it has configured `composer.autoloader_path` - it's not related to this topic), uncomment and complete `$trash->` in `[StupidController.php](https://github.com/przepompownia/example-legacy-project/blob/master/yii2-basic/modules/example/commands/StupidController.php)`.
